File: Chip8Java/src/Chip8.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.awt.event.KeyEvent;

public class Chip8
{
    private final byte[] V = new byte[16];
    private short I;
    private short PC;
    private byte SP = 0;
    private final short[] stack = new short[16];

    private byte DT;
    private byte ST;
    // last time since we clock cycled
    private double lastDecrement = 0;
    // CPU clock speed in Hz
    private final double speed;

    private final byte[] memory = new byte[4096];

    private final String romFname;
    private final Random rand;

    private final int VIDEO_HEIGHT = 32;
    private final int VIDEO_WIDTH = 64;
    private final int[] displayBuffer = new int[VIDEO_WIDTH*VIDEO_HEIGHT];

    private final boolean[] keys = new boolean[16];
    private final Map<Integer, Integer> keymap = new HashMap<>();

    public Chip8(String romName, double clockSpeed)
    {
        romFname = romName;
        rand = new Random();
        speed = clockSpeed;
        init();
    }

    public void init()
    {

        // Setup memory
        byte[] instructions = Loader.readInstructions(romFname);

        System.arraycopy(instructions, 0, memory, 512, instructions.length);

        PC = 512;

        // Add the hexadecimal digits to the interpreter space

        byte[] hexCodes = new byte[]
                {
                        (byte) 0xF0, (byte) 0x90, (byte) 0x90, (byte) 0x90, (byte) 0xF0, // 0
                        (byte) 0x20, (byte) 0x60, (byte) 0x20, (byte) 0x20, (byte) 0x70, // 1
                        (byte) 0xF0, (byte) 0x10, (byte) 0xF0, (byte) 0x80, (byte) 0xF0, // 2
                        (byte) 0xF0, (byte) 0x10, (byte) 0xF0, (byte) 0x10, (byte) 0xF0, // 3
                        (byte) 0x90, (byte) 0x90, (byte) 0xF0, (byte) 0x10, (byte) 0x10, // 4
                        (byte) 0xF0, (byte) 0x80, (byte) 0xF0, (byte) 0x10, (byte) 0xF0, // 5
                        (byte) 0xF0, (byte) 0x80, (byte) 0xF0, (byte) 0x90, (byte) 0xF0, // 6
                        (byte) 0xF0, (byte) 0x10, (byte) 0x20, (byte) 0x40, (byte) 0x40, // 7
                        (byte) 0xF0, (byte) 0x90, (byte) 0xF0, (byte) 0x90, (byte) 0xF0, // 8
                        (byte) 0xF0, (byte) 0x90, (byte) 0xF0, (byte) 0x10, (byte) 0xF0, // 9
                        (byte) 0xF0, (byte) 0x90, (byte) 0xF0, (byte) 0x90, (byte) 0x90, // A
                        (byte) 0xE0, (byte) 0x90, (byte) 0xE0, (byte) 0x90, (byte) 0xE0, // B
                        (byte) 0xF0, (byte) 0x80, (byte) 0x80, (byte) 0x80, (byte) 0xF0, // C
                        (byte) 0xE0, (byte) 0x90, (byte) 0x90, (byte) 0x90, (byte) 0xE0, // D
                        (byte) 0xF0, (byte) 0x80, (byte) 0xF0, (byte) 0x80, (byte) 0xF0, // E
                        (byte) 0xF0, (byte) 0x80, (byte) 0xF0, (byte) 0x80, (byte) 0x80  // F
                };

        System.arraycopy(hexCodes, 0, memory, 0, hexCodes.length);

        // Populate keymap

        keymap.put(KeyEvent.VK_1, 0x1);
        keymap.put(KeyEvent.VK_2, 0x2);
        keymap.put(KeyEvent.VK_3, 0x3);
        keymap.put(KeyEvent.VK_4, 0xC);
        keymap.put(KeyEvent.VK_Q, 0x4);
        keymap.put(KeyEvent.VK_W, 0x5);
        keymap.put(KeyEvent.VK_E, 0x6);
        keymap.put(KeyEvent.VK_R, 0xD);
        keymap.put(KeyEvent.VK_A, 0x7);
        keymap.put(KeyEvent.VK_S, 0x8);
        keymap.put(KeyEvent.VK_D, 0x9);
        keymap.put(KeyEvent.VK_F, 0xE);
        keymap.put(KeyEvent.VK_Z, 0xA);
        keymap.put(KeyEvent.VK_X, 0x0);
        keymap.put(KeyEvent.VK_C, 0xB);
        keymap.put(KeyEvent.VK_V, 0xF);

        // Start an output window

        EmulatorWindow window = new EmulatorWindow(displayBuffer, 640, 320, 64, 32, keymap, keys);
        window.paint(window.getGraphics());

        // Fetch-Decode-Execute Cycle

        while (PC < memory.length-1)
        {
            // making sure that we are running at the correct speed
            double currTime = System.currentTimeMillis();
            double elapsedTime = currTime - lastDecrement;
            if (elapsedTime > (1/speed) * 1000)
            {
                cycle();

                // paint graphics to screen
                window.paint(window.getGraphics());

                lastDecrement = currTime;
            }
        }

        // Exit
        System.out.println("End");
    }

    public void cycle()
    {

            // one fetch decode execute loop

            short ins = fetch();

            PC += 2;

            try
            {
                execute(ins, InsDecode.decodeInstruction(ins));
            }
            catch (UnknownInstructionError e)
            {
                e.printStackTrace();
            }
            // decrement sound and delay timers
            if (DT > 0) DT--;
            if (ST > 0) {ST--; System.out.println("Beep");}
    }

    public short fetch()
    {
        int ins = 0;
        ins = ins | Byte.toUnsignedInt(memory[PC])<<8;
        ins = ins | Byte.toUnsignedInt(memory[PC+1]);
        return (short) ins;
    }

    public void execute(short ins, Instruction type)
    {
        // TO THINK ABOUT - overflows and whatnot - tbh shouldn't be an issue if we got it all stored in appropriate containers
        byte x = InsDecode.getNibble(ins,2);
        byte y = InsDecode.getNibble(ins,1);
        byte kk = InsDecode.getKK(ins);
        short nnn = InsDecode.getNNN(ins);
        switch (type)
        {
            case SYS -> System.out.println();
            case CLS -> Arrays.fill(displayBuffer, 0);
            case RET -> {PC = stack[Byte.toUnsignedInt(SP)]; SP--;}
            case JPn -> PC = nnn;
            // what to do if SP overflows?? ¯\_(ツ)_/¯ programmer's issue
            case CALL -> {SP++; stack[Byte.toUnsignedInt(SP)] = PC; PC = nnn;}
            case SExb -> {if (V[x] == kk) PC += 2;}
            case SNExb -> {if (V[x] != kk) PC += 2;}
            case SExy -> {if (V[x] == V[y]) PC += 2;}
            case LDxb -> V[x] = kk;
            case ADDxb -> V[x] += kk;
            case LDxy -> V[x] = V[y];
            case OR -> V[x] = (byte) (V[x] | V[y]);
            case AND -> V[x] = (byte) (V[x] & V[y]);
            case XOR -> V[x] = (byte) (V[x] ^ V[y]);
            case ADDxy -> {
                if (Byte.toUnsignedInt(V[x]) + Byte.toUnsignedInt(V[y]) > 255)
                {
                    V[0xF] = 1;
                }
                else
                {
                    V[0xF] = 0;
                }
                V[x] += V[y];
            }
            case SUBxy -> {
                if (V[x] > V[y])
                {
                    V[0xF] = 1;
                }
                else
                {
                    V[0xF] = 0;
                }
                V[x] -= V[y];
            }
            case SHR ->
            {
                if ((V[x] & 0b1) == 0b1)
                {
                    V[0xF] = 1;
                }
                else
                {
                    V[0xF] = 0;
                }
                V[x] = (byte) (V[x] >> 1);
            }
            case SUBN -> {
                if (V[y] > V[x])
                {
                    V[0xF] = 1;
                }
                else
                {
                    V[0xF] = 0;
                }
                V[y] -= V[x];
            }
            case SHL ->
            {
                if ((V[x] & -128) == -128)
                {
                    V[0xF] = 1;
                }
                else
                {
                    V[0xF] = 0;
                }
                V[x] = (byte) (V[x] << 1);
            }
            case SNExy -> {if (V[x] != V[y]) PC += 2;}
            case LDI -> I = nnn;
            case JPVn -> PC = (short) (V[0] + nnn);
            case RND -> V[x] = (byte) (rand.nextInt(256) & kk);
            case DRW ->
            {
                byte height = InsDecode.getNibble(ins, 0);
                int xPos = Byte.toUnsignedInt(V[x]) % VIDEO_WIDTH;
                int yPos = Byte.toUnsignedInt(V[y]) % VIDEO_HEIGHT;

                V[0xF] = 0;

                for (int row = 0; row < height; row++)
                {
                    byte spriteRow = memory[I + row];

                    for (int col = 0; col < 8; col++)
                    {
                        byte spritePixel = (byte)(spriteRow & (0x80 >> col));

                        // make sure to mod it so it wraps around
                        int bufferPos = ((yPos + row) * VIDEO_WIDTH + (xPos + col)) % displayBuffer.length;

                        int screenPixel = displayBuffer[bufferPos];

                        if (Byte.toUnsignedInt(spritePixel) > 0)
                        {
                            if (screenPixel == 1)
                            {
                                V[0xF] = 1;
                            }
                            displayBuffer[bufferPos] ^= 1;
                        }
                    }
                }
            }
            case SKP -> {if (keys[Byte.toUnsignedInt(V[x])]) PC += 2;}
            case SKNP -> {if (!keys[Byte.toUnsignedInt(V[x])]) PC += 2;}
            case LDxDT -> V[x] = DT;
            case LDxK -> {
                boolean keyPress = false;
                for (int i = 0; i < keys.length; i++)
                {
                    if (keys[i])
                    {
                        V[x] = (byte) i;
                        keyPress = true;
                    }
                }
                if (!keyPress)
                {
                    PC -= 2;
                }
            }
            case LDDTx -> DT = V[x];
            case LDST -> ST = V[x];
            case ADDI -> I += V[x];
            case LDF -> I = (short)(V[x] * 5);
            case LDB ->
            {
                memory[I] = (byte) ((Byte.toUnsignedInt(V[x]) / 100) % 10);
                memory[I+1] = (byte) ((Byte.toUnsignedInt(V[x]) / 10) % 10);
                memory[I+2] = (byte) (Byte.toUnsignedInt(V[x]) % 10);
            }
            case LDIx ->
            {
                if (x + 1 >= 0) System.arraycopy(V, 0, memory, I, x + 1);
            }
            case LDxI ->
            {
                if (x + 1 >= 0) System.arraycopy(memory, I, V, 0, x + 1);
            }
        }
    }
}

File: Chip8C/src/chip8.c
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <time.h>

#include <instructions.h>
#include <chip8.h>

int decode_ins(unsigned short ins);

unsigned char memory[MEM_SIZE] = {0};
unsigned char V[16] = {0};
unsigned short I;
unsigned short PC;
unsigned char SP = 0;
unsigned short stack[16] = {0};

unsigned char DT;
unsigned char ST;
    
// last time since we clock cycled
float last_decrement = 0;
// CPU clock speed in Hz
float speed;

bool display_buf[VIDEO_WIDTH * VIDEO_HEIGHT] = {0};

bool keys[16] = {0};

int execute(unsigned short ins)
{
    unsigned char x = (ins & 0xF00) >> 8;
    unsigned char y = (ins & 0xF0) >> 4;
    unsigned char kk = ins & 0xFF;
    unsigned short nnn = ins & 0xFFF;
    
    switch(decode_ins(ins))
    {
        case CLS:
            // Clear the display
            memset(memory, 0, MEM_SIZE);
            break;
        case RET:
            // Return from a subroutine
            PC = stack[SP--];
            break;
        case SYS:
            // Jump to a machine code routine at nnn
            // do nothing pls
            break;
        case JPn:
            // Jump to location nnn
            PC = nnn;
            break;
        case CALL:
            // Call subroutine at nnn
            stack[++SP] = PC;
            PC = nnn;
            break;
        case SExb:
            // Skip next instruction if Vx == kk
            if (V[x] == kk) PC += 2;
            break;
        case SNExb:
            // Skip next instruction if Vx != kk
            if (V[x] != kk) PC += 2;
            break;
        case SExy:
            // Skip next instruction if Vx == Vy
            if (V[x] == V[y]) PC += 2;
            break;
        case LDxb:
            // Set Vx = kk
            V[x] = kk;
            break;
        case ADDxb:
            // Set Vx = Vx + kk
            V[x] += kk;
            break;
        case LDxy:
            // Set Vx = Vy
            V[x] = V[y];
            break;
        case OR:
            // Set Vx = Vx OR Vy
            V[x] |= V[y];
            break;
        case AND:
            // Set Vx = Vx AND Vy
            V[x] &= V[y];
            break;
        case XOR:
            // Set Vx = Vx XOR Vy
            V[x] ^= V[y];
            break;
        case ADDxy:
            // Set Vx = Vx + Vy, set VF = carry
            V[x] += V[y];
            V[0xF] = ((int)V[x] + (int)V[y]) > 255 ? 1 : 0;
            break;
        case SUBxy:
            // Set Vx = Vx - Vy, set VF = NOT borrow
            V[0xF] = V[x] > V[y] ? 1 : 0;
            V[x] -= V[y];
            break;
        case SHR:
            // Set Vx = Vx SHR 1
            V[0xF] = V[x] & 1;
            V[x] >>= 1;
            break;
        case SUBN:
            // Set Vx = Vy - Vx, set VF = NOT borrow
            V[0xF] = V[y] > V[x] ? 1 : 0;
            V[x] = V[y] - V[x];
            break;
        case SHL:
            // Set Vx = Vx SHL 1
            V[0xF] = V[x] >> 7;
            V[x] <<= 1;
            break;
        case SNExy:
            // Skip next instruction if Vx != Vy
            if (V[x] != V[y]) PC += 2;
            break;
        case LDI:
            // Set I = nnn
            I = nnn;
            break;
        case JPVn:
            // Jump to location nnn + V0
            PC = nnn + V[0];
            break;
        case RND:
            // Set Vx = random byte AND kk
            V[x] = ((unsigned char) rand()) & kk;
            break;
        case DRW:
            // Display n-byte sprite starting at memory location I at (Vx, Vy), set VF = collision
            unsigned char height = ins & 0xF;
            int x_pos = V[x] % VIDEO_WIDTH;
            int y_pos = V[y] % VIDEO_HEIGHT;

            V[0xF] = 0;

            for (int row = 0; row < height; row++)
            {
                unsigned char sprite_row = memory[I+row];
                for (int col = 0; col < 8; col++)
                {
                    unsigned char sprite_pixel = sprite_row & (0x80 >> col);

                    // make sure to mod it so it wraps around
                    int buffer_pos = ((y_pos + row) * VIDEO_WIDTH + (x_pos + col)) % (VIDEO_HEIGHT*VIDEO_WIDTH);

                    int screen_pixel = display_buf[buffer_pos];

                    if(sprite_pixel > 0)
                    {
                        if (screen_pixel == 1)
                        {
                            V[0xF] = 1;
                        }
                        display_buf[buffer_pos] ^= 1;
                    }
                }
            }

            break;
        case SKNP:
            // Skip next instruction if key with the value of Vx is not pressed
            if (keys[V[x]]) PC += 2;
            break;
        case SKP:
            // Skip next instruction if key with the value of Vx is pressed
            if (!keys[V[x]]) PC += 2;
            break;
        case LDxDT:
            // Set Vx = delay timer value
            V[x] = DT;
            break;
        case LDxK:
            // Wait for a key press, store the value of the key in Vx
            bool key_press = false;
            for (unsigned char i = 0; i < 16; i++)
            {
                if (keys[i])
                {
                    V[x] = i;
                    key_press = true;
                }
            }
            if (!key_press)
            {
                PC -= 2;
            }
            break;
        case LDDTx:
            // Set delay timer = Vx
            DT = V[x];
            break;
        case LDST:
            // Set sound timer = Vx
            ST = V[x];
            break;
        case ADDI:
            // Set I = I + Vx
            I += V[x];
            break;
        case LDF:
            // Set I = location of sprite for digit Vx
            I = V[x]*5;
            break;
        case LDB:
            // Store BCD representation of Vx in memory locations I, I+1, and I+2
            memory[I] = (V[x]/100)%10;
            memory[I+1] = (V[x]/10)%10;
            memory[I+2] = V[x]%10;
            break;
        case LDIx:
            // Store registers V0 through Vx in memory starting at location I
            memcpy(memory + I, V, x+1);
            break;
        case LDxI:
            // Read registers V0 through Vx from memory starting at location I
            memcpy(V, memory + I, x+1);
            break;
        default:
            // Handle unknown instruction
            printf("Unknown instruction 0x%x\n", ins);
            return -1;
    }
    return 0;
}

/**
 * Take a unsigned short instruction and decode it to an integer macro code
 */
int decode_ins(unsigned short ins)
{
    unsigned char first = (ins & 0xF000) >> 12;
    
    switch(first)
    {
        case 0:
            switch (ins & 0xFFF)
            {
                case 0x0E0:
                    return CLS;
                case 0x0EE:
                    return RET;
                default:
                    return SYS;
            }

        case 1:
            return JPn;
        case 2:
            return CALL;
        case 3:
            return SExb;
        case 4:
            return SNExb;
        case 5:
            return SExy;
        case 6:
            return LDxb;
        case 7:
            return ADDxb;

        case 8:
            switch(ins & 0xF)
            {
                case 0:
                    return LDxy;
                case 1:
                    return OR;
                case 2:
                    return AND;
                case 3:
                    return XOR;
                case 4:
                    return ADDxy;
                case 5:
                    return SUBxy;
                case 6:
                    return SHR;
                case 7:
                    return SUBN;
                case 0xE:
                    return SHL;
                default:
                    return -1;
            }

        case 9:
            return SNExy;
        case 0xA:
            return LDI;
        case 0xB:
            return JPVn;
        case 0xC:
            return RND;
        case 0xD:
            return DRW;
        
        case 0xE:
            switch (ins & 0xF)
            {
                case 1:
                    return SKNP;
                case 2:
                    return SKP;
                default:
                    return -1;
            }

        case 0xF:
            switch (ins & 0xFF)
            {
                case 0x7:
                    return LDxDT;
                case 0x0A:
                    return LDxK;
                case 0x15:
                    return LDDTx;
                case 0x18:
                    return LDST;
                case 0x1E:
                    return ADDI;
                case 0x29:
                    return LDF;
                case 0x33:
                    return LDB;
                case 0x55:
                    return LDI;
                case 0x65:
                    return LDxI;
                default:
                    return -1;
            }
        default:
            return -1;
    }
    return -1;
}

unsigned short fetch()
{
    unsigned short ins = 0;
    ins |= memory[PC] << 8;
    ins |= memory[PC+1];
    return ins;
}

void cycle()
{
    // one FDE loop

    unsigned short ins = fetch();
    PC += 2;
    
    if (execute(ins) == -1)
    {
        printf("Error occurred.\n");
        return;
    }

    //decrement sound and delay timers

    if (DT > 0) DT--;
    if (ST > 0) {ST--; printf("Beep.\n");}
}

void init()
{
    PC = 512;

    //load hex digit sprites into interpreter space

    unsigned char hex_codes[80] = 
    {
        0xF0,  0x90,  0x90,  0x90,  0xF0, // 0
        0x20,  0x60,  0x20,  0x20,  0x70, // 1
        0xF0,  0x10,  0xF0,  0x80,  0xF0, // 2
        0xF0,  0x10,  0xF0,  0x10,  0xF0, // 3
        0x90,  0x90,  0xF0,  0x10,  0x10, // 4
        0xF0,  0x80,  0xF0,  0x10,  0xF0, // 5
        0xF0,  0x80,  0xF0,  0x90,  0xF0, // 6
        0xF0,  0x10,  0x20,  0x40,  0x40, // 7
        0xF0,  0x90,  0xF0,  0x90,  0xF0, // 8
        0xF0,  0x90,  0xF0,  0x10,  0xF0, // 9
        0xF0,  0x90,  0xF0,  0x90,  0x90, // A
        0xE0,  0x90,  0xE0,  0x90,  0xE0, // B
        0xF0,  0x80,  0x80,  0x80,  0xF0, // C
        0xE0,  0x90,  0x90,  0x90,  0xE0, // D
        0xF0,  0x80,  0xF0,  0x80,  0xF0, // E
        0xF0,  0x80,  0xF0,  0x80,  0x80  // F
    };

    memcpy(memory, hex_codes, 80);

    // start an output window

    // start input handler

    // FDE Cycle

    while (PC < MEM_SIZE-1)
    {
        cycle();
        nanosleep(&ts, &ts);
    }

    printf("END\n");
}